test/unit/core/simd/simd_algorithm_test.cpp

```diff
@@ -38,3 +41,236 @@ TEST(simd_algorithm, iota)
+        SIMD_EQ(trd, vec);
+        SIMD_EQ(fou, vec);
+        SIMD_EQ(fif, vec);
+        SIMD_EQ(six, vec);
+        SIMD_EQ(sev, vec);
+        SIMD_EQ(eig, vec);
+    }
+    else
+    {
+        for (size_t idx = 0; idx < simd_traits<TypeParam>::length / 8; ++idx)
+        {
+            EXPECT_EQ(fst[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 0]);
+            EXPECT_EQ(sec[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 1]);
+            EXPECT_EQ(trd[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 2]);
+            EXPECT_EQ(fou[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 3]);
+            EXPECT_EQ(fif[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 4]);
+            EXPECT_EQ(six[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 5]);
+            EXPECT_EQ(sev[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 6]);
+            EXPECT_EQ(eig[idx], vec[idx + simd_traits<TypeParam>::length / 8 * 7]);
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------
+// Algorithm upcast
+//-----------------------------------------------------------------------------
+
+template <typename t>
+class simd_algorithm_upcast : public ::testing::Test
+{
+public:
+
+    static constexpr auto target_list_signed()
+    {
+        if constexpr (sizeof(t) == 1)
+            return type_list_signed_8{};
+        else if constexpr (sizeof(t) == 2)
+            return type_list_signed_16{};
+        else
+            return type_list_signed_32{};
+    }
+
+    static constexpr auto target_list_unsigned()
+    {
+        if constexpr (sizeof(t) == 1)
+            return type_list_unsigned_8{};
+        else if constexpr (sizeof(t) == 2)
+            return type_list_unsigned_16{};
+        else
+            return type_list_unsigned_32{};
+    }
+
+    using type_list_signed_8  = type_list<int16_t, int32_t, int64_t>;
+    using type_list_signed_16 = type_list<int32_t, int64_t>;
+    using type_list_signed_32 = type_list<int64_t>;
+
+    using type_list_unsigned_8  = type_list<uint16_t, uint32_t, uint64_t>;
+    using type_list_unsigned_16 = type_list<uint32_t, uint64_t>;
+    using type_list_unsigned_32 = type_list<uint64_t>;
+};
+
+using upcast_test_types = ::testing::Types<int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t>;
+TYPED_TEST_CASE(simd_algorithm_upcast, upcast_test_types);
+
+TYPED_TEST(simd_algorithm_upcast, signed)
+{
+    using list = decltype(TestFixture::target_list_signed());
+
+    detail::for_each_type([this] (auto type_id)
+    {
+        using target_type = typename decltype(type_id)::type;
+        using src_simd_t = simd_type_t<TypeParam>;
+        using tar_simd_t = simd_type_t<target_type>;
+
+        src_simd_t s = fill<src_simd_t>(-10);
+        tar_simd_t t = upcast<tar_simd_t>(s);
+
+        for (size_t i  = 0; i < simd_traits<tar_simd_t>::length; ++i)
+            EXPECT_EQ(t[i], static_cast<target_type>(static_cast<TypeParam>(-10)));
```

Comment:
not sure what happened here 🤔

Reply:
Ah because it needs to be this way. SIMD_EQ wants the same saclar types but I want to explicitly test the upcasting is done correctly also in the signed case. So I need to do it element wise.

Reply:
can you add a comment to make that clear?